### PR TITLE
Support binding for slice/array obj [Rewrite]

### PR DESCRIPTION
**Pull request by wuhuizuo**
support usage followed:

```golang
type StructExample struct {
    A string `json:"a" form:"a" binding"required,max=16"`
}

func exampleHandler(c *gin.Context) {
    var data []StructExample

    if c.Bind(&data) != nil {
        return
    }
    c.JSON(http.StatusOK, data)
}
```

**this pull is rewritten for https://github.com/gin-gonic/gin/pull/2194， avoid to change the validator interface**

### CODE REVIEW

**Comment by wuhuizuo** [comment]
@thinkerou @appleboy I have rewritten for review issues in !2194

File: binding/default_validator_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// license that can be found in the LICENSE file.
+
+package binding
+
+import (
+	"errors"
+	"testing"
+)
+
+func TestSliceValidateError_Error(t *testing.T) {
```
Comment (thinkerou) [inline binding/default_validator_test.go:12]: please rename using `TestSliceValidateError`

File: binding/default_validator_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.err.Error(); got != tt.want {
+				t.Errorf("sliceValidateError.Error() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_defaultValidator_Validate(t *testing.T) {
```
Comment (thinkerou) [inline binding/default_validator_test.go:29]: please also rename, thanks!

File: binding/default_validator.go [hunk 3]
```diff
@@ -16,22 +18,57 @@ type defaultValidator struct {
...
 	if valueType == reflect.Ptr {
-		valueType = value.Elem().Kind()
+		return v.ValidateStruct(value.Elem().Interface())
 	}
-	if valueType == reflect.Struct {
-		v.lazyinit()
-		if err := v.validate.Struct(obj); err != nil {
-			return err
+
+	switch valueType {
```
Comment (thinkerou) [inline binding/default_validator.go:48]: use
```
switch value.Kind() {
    case reflect.Ptr:
...
```

File: binding/default_validator_test.go (hunk 1 above)
Comment (wuhuizuo) [reply binding/default_validator_test.go]: fixed in 8098d88562163a827f495e72b49bb936f0ebe881

File: binding/default_validator_test.go (hunk 2 above)
Comment (wuhuizuo) [reply binding/default_validator_test.go]: > please also rename, thanks!

fixed in 8098d88562163a827f495e72b49bb936f0ebe881

File: binding/default_validator.go (hunk 3 above)
Comment (wuhuizuo) [reply binding/default_validator.go]: > use
> 
> ```
> switch value.Kind() {
>     case reflect.Ptr:
> ...
> ```

fixed in c29edbb8a059ba8c0271f6f45e3813dbf03b2aba

File: binding/default_validator_test.go (hunk 1 above)
Comment (wuhuizuo) [reply binding/default_validator_test.go]: > please rename using `TestSliceValidateError`

fixed in 03ce1a6b864de65c92f305e8b6d90c1675f2c9e2

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
@wuhuizuo Please fix the conflicts.

**Comment by wuhuizuo** [comment]
> @wuhuizuo Please fix the conflicts.

done.

**Comment by appleboy** [comment]
@wuhuizuo build fail: https://travis-ci.org/github/gin-gonic/gin/builds/740820459

File: binding/binding_test.go [hunk 4]
```diff
@@ -1181,6 +1195,21 @@ func testBodyBinding(t *testing.T, b Binding, name, path, badPath, body, badBody
 	assert.Error(t, err)
 }
 
+feature/array_support_for_json_bind
```
Comment (thinkerou) [inline binding/binding_test.go:1198]: SYNTAX ERROR

Comment (wuhuizuo) [reply binding/binding_test.go]: sorry! it's fixed.

**Review by thinkerou — APPROVED** [review]
lgtm

**Comment by appleboy** [comment]
LGTM but build fail in go master version. See the report: https://travis-ci.org/github/gin-gonic/gin/jobs/742947069

**Comment by thinkerou** [comment]
@appleboy related issue: #2553

**Review by appleboy — APPROVED** [review]

**Pull request merged**